docs/polkadot/implementors-guide/guide.md

```diff
@@ -994,9 +992,44 @@ Dispatch a `PovFetchSubsystemMessage(relay_parent, candidate_hash, sender)` and
 
 (TODO: send statements to Statement Distribution subsystem, handle shutdown signal from candidate backing subsystem)
 
+### Candidate Selection Subsystem
+
+#### Description
+
+The Candidate Selection subsystem monitors net traffic for two events:
+
+- a new parablock candidate is available
+- a peer has seconded a parablock candidate
+
+This module is only ever interested in parablocks assigned to the particular parachain which this validator is currently handling.
+
+New parablock candidates may arrive from a potentially unbounded set of collators. This subsystem chooses either 0 or 1 of them per relay parent to second. If it chooses to second a candidate, it sends an apropriate message to the **Candidate Backing** subsystem to generate an appropriate `Statement`.
+
+All parablocks which peers have seconded are also sent to the Candidate Backing subsystem for re-validation.
+As seconded peers are tallied, double-votes are detected. If found, a report is sent to the **Misbehavior Arbitration** subsystem.
```

Comment:
There is a division to be clarified about which subsystems are responsible for double-vote reporting. Candidate Backing, Statement Distribution, and Candidate Seconding subsystems are all going to be in position to detect double-vote misbehavior. We shouldn't be filing duplicate reports from all systems, since one should be sufficient. Of these, I think Statement Distribution is probably the best fit, as it's also the system that's capable of dropping new incoming statements - the marginal cost of double-voting more than once is 0, so someone who double-votes is likely to spam.

We could establish an invariant that statement distribution doesn't give other subsystems double-votes.

Reply:
OK, I don't feel strongly about which of those subsystems has the formal responsibility to detect double-votes. Agree that only one of them should really be in charge of it.

Not sure how to establish that invariant: couldn't we get Vote 1, distribute it, then receive Vote 2 from the same source? We can report double-voting at that point, and we can cease to distribute Vote 1, but we can't rescind those instances we've already sent.

Reply:
What's the distinction between Candidate Backing and Candidate Seconding, by the way? The latter is new to me.

Reply:
We do not afaik require double vote defenses:  

We should prevent a parablocks' first backer from being first backer for another parablock until it either progresses to availability or expires, but that's only so that one bad first backer cannot halt a parachain for one rotation, and the logic appears easy since replacing an expired candidate receipt reveals the previous first backer.  

We could encourage duplicate backing votes on competing parablock because the relay chain block producer chooses among them definitively.  Availability and approval votes cannot be duplicates by construction.

Reply:
What does it mean to be the "first backer"? Ordering is really relative to network conditions. This first-backer constraint sounds complicated and don't seem to get us much, as the para will still be live upon the next rotation. Double-votes are the only misbehavior we want to want to discourage at this phase as everything else is covered by the approval system, which hasn't been specified yet. It's easy to avoid double-voting as an honest validator and the slashing condition is strict.

Reply:
BTW, by double-voting I've meant one of two things:
  - Issuing two `Seconded` statements.
  - Issuing conflicting statements on the same candidate (Valid/Invalid, Seconded/Valid, Seconded/Invalid)

Reply:
Yes exactly.  I never even considered conflicting votes on the same candidate, but yes those are slashable.

Reply:
So, we agree that one validator `Second`ing two different parablocks with the same parents is a double-vote and should be slashed.

What I am still confused by: is it an error if two different validators both `Second` the same parablock? If yes: why, exactly; if there's a failure there, it's not obvious to me. If no: could we consider removing the `Seconded` variant from `Statement` and passing `CandidateReceipt`s via some other channel?

Reply:
No.  A&V always extracts consensus from the relay chain to avoid engaging in its own BFT protocol within parachain validator groups. 

We want several parachain validators to second each candidate, perhaps 2/3rds, so if more than 1/3rd second two candidates then the parachain halts unless they can second more choices.

We can replace my suggestion to rotate "first backers" by merely asking that anytime we replace an expired candidate with signer set `S_old` by a new candidate with signer set `S_new` then `S_new` must include `min_new_backers` validators not in `S_old`.  We'd take `min_new_backers=0` if we asked for 2/3rds, but some parameter choices give `min_new_backers>0`.  We could then second as many parablock with the same parent as we like.

Reply:
Thanks! In that case, I'd propose to simplify `Statement`s:

```rust
/// A statement about the validity of a parachain candidate
enum Statement {
    /// This parachain candidate is valid.
    Valid(CandidateReceipt),
    /// The parachain candidate with this hash is invalid.
    Invalid(Hash),
}
```

Motivation: eliminating the `Seconded` variant makes it much more clear what the semantics should be: a validator should never contradict its own statements about a block, and all validators should agree on the semantics. As far as I can tell, the only motivation for distinguishing the `Seconded` variant is to save some bytes for those validators who mark a block `Valid`. That feels like premature optimization, to me.

Reply:
Yup, we forbid self conflicts.  :)  

I still think XCMP might distinguish between the first backer and later backers, meaning multiple parachain validators could compute different candidate receipts from the same parachain block, so "seconding" could refer to backing parachain validators after the first one.  I'd assumed you meant this initially, but we should avoid assuming a deterministic map from candidates to candidate receipts.

Reply:
> As far as I can tell, the only motivation for distinguishing the Seconded variant is to save some bytes for those validators who mark a block Valid. That feels like premature optimization, to me.

Actually, it's to bound the number of candidates needed to validate, which is a legitimate DoS vector. The `Seconded` message gives us an upper bound on the number of candidates we'd ever need to validate or distribute during the backing phase, as long as nobody wants to get slashed 100%.

> We can replace my suggestion to rotate "first backers" by merely asking that anytime we replace an expired candidate with signer set S_old by a new candidate with signer set S_new then S_new must include min_new_backers validators not in S_old. We'd take min_new_backers=0 if we asked for 2/3rds, but some parameter choices give min_new_backers>0. We could then second as many parablock with the same parent as we like.

I'm still not sure why we want this - does it actually give us any more security, given that we will ensure the block is available and we have a secondary-checking system.

Reply:
@burdges and I talked separately, the issues are based on different mental models and are resolved by having collators fully responsible for generating PoVs, as they do in this document, so the first backer is not as important. We should keep the `Seconded` message and don't need to worry about the candidate-replacement signer overlap stuff as with this division between collation and validation it is a liveness concern as opposed to a safety one.

Reply:
We figured this out.  We'll want a clearer deliniation between collation and backing than my notion of "first" gave.  It does not preclude later running the last phase of collation on validators, but this becomes an architectural question for cumulus, not A&V.  :)